src/main/scala/mesosphere/marathon/api/v2/TasksResource.scala

```diff
@@ -30,22 +30,44 @@ class TasksResource @Inject() (
   @GET
   @Produces(Array(MediaType.APPLICATION_JSON))
   @Timed
-  def indexJson(@QueryParam("status") status: String,
-                @QueryParam("status[]") statuses: util.List[String]): Response = {
+  def indexJson(
+    @QueryParam("status") status: String,
+    @QueryParam("status[]") statuses: util.List[String],
+    @QueryParam("limit")@DefaultValue("-1") limit: Int,
+    @QueryParam("offset")@DefaultValue("0") offset: Int): Response = {
     if (status != null) statuses.add(status)
     val statusSet = statuses.asScala.flatMap(toTaskState).toSet
+
+    val tasks = taskTracker.list.values.iterator.flatMap { app =>
+      app.tasks.iterator.map(t => app.appName -> t)
+    }
+
+    val appToPorts = taskTracker.list.keySet.map { appId =>
+      appId -> service.getApp(appId).map(_.servicePorts).getOrElse(Nil)
+    }.toMap
+
+    val health = taskTracker.list.keySet.flatMap { appId =>
+      result(healthCheckManager.statuses(appId))
+    }.toMap
+
+    val enrichedTasks = for {
+      (appId, task) <- tasks
+      if statusSet.isEmpty || statusSet(task.getStatus.getState)
+    } yield {
+      EnrichedTask(
+        appId,
+        task,
+        health.getOrElse(task.getId, Nil).map(Some(_)),
+        appToPorts(appId)
+      )
+    }
+
     ok(Map(
-      "tasks" -> taskTracker.list.flatMap {
-        case (appId, setOfTasks) =>
-          setOfTasks.tasks.collect {
-            case task if statusSet.isEmpty || statusSet(task.getStatus.getState) =>
-              EnrichedTask(
-                appId,
-                task,
-                result(healthCheckManager.status(appId, task.getId)),
-                service.getApp(appId).map(_.servicePorts).getOrElse(Nil)
-              )
-          }
+      "tasks" -> {
+        if (limit >= 0)
+          enrichedTasks.drop(offset).take(limit)
```

Comment:
It also feels fragile to me, to return a mutable value from such a method (Iterator). How do I ensure that it is only used once by the caller?

If it is on purpose, because of framework magic, please explain.